### PR TITLE
Add Handling of Cache Prompt When Formatting Messages

### PR DESCRIPTION
**Pull request by ColeMurray**
**Short description of the problem this fixes or functionality that this introduces. This may be used for the CHANGELOG**
Implements correct application of cache control headers in message formatting based on the cache_prompt flag.

**Give a summary of what the PR does, explaining any non-trivial design decisions**
This PR addresses an issue where the cache control header was not being correctly applied when formatting messages. The changes include:

1. Modified the `format_messages` function to track the `cache_prompt` flag for each content item in a message.
2. Implemented logic to combine `cache_prompt` flags from multiple content items using the OR operator.
3. Added a condition to include the `cache_control` key with the value 'ephemeral' in the formatted message when `cache_prompt` is True.
4. Updated the handling of different message types (string, dict, Message object) to properly check and apply the `cache_prompt` flag.

These changes ensure that the cache control header is correctly applied based on the `cache_prompt` flag of the content items, improving the caching behavior of the system.

---
**Link of any specific issues this addresses**
Fixes https://github.com/All-Hands-AI/OpenHands/issues/3772: Incorrect application of cache control headers in message formatting

### CODE REVIEW

File: openhands/core/message.py
Comment (enyst) [inline openhands/core/message.py:114]:
```suggestion
            formatted_message: dict[str, Any] = {
```

Comment (enyst) [reply openhands/core/message.py:114]: We use python 3.11+, so we can use the built-in, simpler version.

**Comment by tobitege** [comment]
@ColeMurray good catch! I forgot that important part. Thanks for the PR!

File: openhands/core/message.py
Comment (enyst) [inline openhands/core/message.py:119]: Can you please help me understand this, we are setting 'cache_control' on a message, not on a TextContext? I expected we need to do it on the TextContent or ImageContent, I seem to recall that the LLM will indeed respect it per item.

Comment (enyst) [reply openhands/core/message.py:119]: What do you think about this: these are pydantic model classes, and they have each their own @model_serializer. When we used `model_dump` on the Message, it applied it to its component items, and we got the right format for free, so to speak. Now there seem to be some reasons why we don't rely on it anymore for the Messages, but could we still use it on TextContent and respectively ImageContent? It sets the appropriate cache_control if the `cache_prompt` attribute is set, I think?

Comment (tobitege) [reply openhands/core/message.py:119]: The code change is happening in a conversion method, which gets rid of the pydantic structure (like TextContent and  an extra type attribute), as that is an issue with several models, that error on that.

Comment (enyst) [reply openhands/core/message.py:119]: @tobitege you mean that this [`type`](https://github.com/All-Hands-AI/OpenHands/blob/5100d12cea2cd35c30a22e25fbac376b72ed0981/openhands/core/message.py#L31) shouldn't be in the dict when vision is disabled or unsupported, but should be when vision is enabled and supported? Ugh. Then scratch that, it does affect TextContent.

I'd still appreciate an answer to the first question. We do something a little unusual [here](https://github.com/All-Hands-AI/OpenHands/blob/5100d12cea2cd35c30a22e25fbac376b72ed0981/agenthub/codeact_agent/codeact_agent.py#L289). This TextContent and thus the last Message it's part of, change next step, when the new message added to the list is now the last message and gets the reminder. I _think_ the way it worked with prompt caching is when the cache was set on the TextContent _before_ this last one.

Comment (tobitege) [reply openhands/core/message.py:119]: For e.g. Gemini (with vision) throws an error that the system message doesn't allow any extra attributes, which would be the `type` one, similarly it happens with models on Groq's API, iirc.

Yes, you're correct, we still have this `cache_prompt` on `TextContent`-level elsewhere:
https://github.com/All-Hands-AI/OpenHands/blob/5100d12cea2cd35c30a22e25fbac376b72ed0981/agenthub/codeact_agent/codeact_agent.py#L227-L238

Within this method here, the extra environmental hint would become part of a single-string (`messages` items on top) and not be its own message. 
Correct me if I'm wrong, but if vision were _enabled_ the problem would be with the pydantic serialization, not with the method here actually? 🤔

**Comment by tobitege** [comment]
@ColeMurray Thanks for the contribution.
In your dev environment, you may look into the `dev_config/python` folder and how to install the `pre-commit-config`.
That would run ruff and mypy upon commit for you and fix most end-of-file issues automatically. 
Could save you some time and hassle to keep the workflows happy. 😃

**Comment by ColeMurray** [comment]
> @ColeMurray Thanks for the contribution. In your dev environment, you may look into the `dev_config/python` folder and how to install the `pre-commit-config`. That would run ruff and mypy upon commit for you and fix most end-of-file issues automatically. Could save you some time and hassle to keep the workflows happy. 😃

The specific lint error was introduced by accepting the change here, which missed the import https://github.com/All-Hands-AI/OpenHands/pull/3773#discussion_r1749209458

I've pushed up a fix

**Comment by tobitege** [comment]
> The specific lint error was introduced by accepting the change here, which missed the unused import [#3773 (comment)](https://github.com/All-Hands-AI/OpenHands/pull/3773#discussion_r1749209458)
> 
> I've pushed up a fix

Yes, no worries. Haven't looked into this if we can fix that too someday. :)

**Comment by tobitege** [comment]
I'll defer to @enyst for the final review in light of the previous discussion.

File: openhands/core/message.py
Comment (enyst) [reply openhands/core/message.py:119]: Well, the CodeAct code on prompt caching worked in the past, using `model_dump()`.

> Within this method here, the extra environmental hint would become part of a single-string (messages items on top) and not be its own message.

No, it's its own TextContent object, added to a Message that has one other TextContent before it. So the result of that method you linked is: a Message with 2 TextContent objects, the first has `cache_prompt` and should be serialized with `cache_control`, the last is the reminder and should not have cache.

Why are we merging all TextContents in a single string? Sorry if I'm missing something, can't we send the dicts or something corresponding to the TextContent objects?

Comment (enyst) [reply openhands/core/message.py:119]: Tobi, I'm a little confused here too:
> For e.g. Gemini (with vision) throws an error that the system message doesn't allow any extra attributes, which would be the `type` one, similarly it happens with models on Groq's API, iirc.

But as you said, if vision is enabled this method is not executed further, in this file at [line 80](https://github.com/All-Hands-AI/OpenHands/pull/3773/files#diff-1cdab1e3cbde6aeaac166359980171d0fe0c4c85688b5fe82cd38e249969cf23L80) it falls back to pydantic. So how did this method solve the problem of `type` for Gemini with vision? Or did you mean without vision? 🤔

**Comment by enyst** [comment]
@ColeMurray Thank you for the work on this! And sorry for the time taken with discussions.

With the change in the other PR, could you please confirm that prompt caching for Anthropic still doesn't work for you?

In that case, could you enable vision in config (for example run with LLM_DISABLE_VISION=0), and confirm that it still doesn't work?

**Comment by ColeMurray** [comment]
> @ColeMurray Thank you for the work on this! And sorry for the time taken with discussions.
> 
> With the change in the other PR, could you please confirm that prompt caching for Anthropic still doesn't work for you?
> 
> In that case, could you enable vision in config (for example run with LLM_DISABLE_VISION=0), and confirm that it still doesn't work?

Confirm this does not work without this change.

Steps to repro:
1. Pull latest from main
2. ```LLM_DISABLE_VISION=0 DEBUG=1 make run```
3. Observe in logs no cache tokens being written
```
17:47:45 - openhands:INFO: llm.py:528 - Cost: 0.02 USD | Accumulated Cost: 0.03 USD
Input tokens: 5103
Output tokens: 70

17:47:45 - openhands:DEBUG: agent_controller.py:238 - [Agent Controller d53ff0ac-9e16-4149-a5cf-475f412d4e76] Setting agent(CodeActAgent) state from AgentState.RUNNING to AgentState.AWAITING_USER_CONFIRMATION
```

This failure matches my expectations, as without this change the result of the ```format_messages_for_llm``` will not have the cache_control property which is needed for prompt caching
https://github.com/All-Hands-AI/OpenHands/blob/main/agenthub/codeact_agent/codeact_agent.py#L197
https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/core/message.py#L74

https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching

File: openhands/core/message.py
Comment (tobitege) [reply openhands/core/message.py:119]: > So how did this method solve the problem of `type` for Gemini with vision? Or did you mean without vision? 🤔

Sorry, that is a typo indeed, I meant "without".

**Comment by enyst** [comment]
@ColeMurray Strange, I just ran with vision enabled, and it worked without this PR's changes, because it exits the method [here](https://github.com/All-Hands-AI/OpenHands/pull/3773/files#diff-1cdab1e3cbde6aeaac166359980171d0fe0c4c85688b5fe82cd38e249969cf23L81). So it never executes your additions.

Logs:
> 20:33:32 - openhands:INFO: llm.py:528 - Cost: 0.03 USD | Accumulated Cost: 0.10 USD
Input tokens: 31
Output tokens: 301
Input tokens (cache write): 5048
Input tokens (cache read): 9191


> This failure matches my expectations, as without this change the result of the `format_messages_for_llm` will not have the cache_control property which is needed for prompt caching

But it does have it, from `[message.model_dump() for message in messages]` code. This is the last message, 2 TextContent dicts:
Screenshot 2024-09-09 at 20 19 02

So it must be that for you it didn't have `with_images` true at line 80. I'll have to try to replicate that. 🤔  You chose the model from the UI, and had Anthropic prefix, right? Which model exactly are you using?

**Comment by enyst** [comment]
OK, I can reproduce this when I run with frontend: frontend adds "anthropic/" to the model name, so it becomes "anthropic/claude-...". But LiteLLM `model_cost` doesn't contain "anthropic/" pattern, so liteLLM `supports_vision` returns false. So vision is not enabled.

Cc: @amanape do you know why we use "anthropic/" with a slash? Can we use simple "claude-stuff" instead? Everything (vision, prompt cache) works with "claude..." names, but not with "anthropic/claude..." names, because litellm doesn't recognize it as vision-enabled. 🤔

**Comment by ColeMurray** [comment]
> @ColeMurray Strange, I just ran with vision enabled, and it worked without this PR's changes, because it exits the method [here](https://github.com/All-Hands-AI/OpenHands/pull/3773/files#diff-1cdab1e3cbde6aeaac166359980171d0fe0c4c85688b5fe82cd38e249969cf23L81). So it never executes your additions.
> 
> Logs:
> 
> > 20:33:32 - openhands:INFO: llm.py:528 - Cost: 0.03 USD | Accumulated Cost: 0.10 USD
> > Input tokens: 31
> > Output tokens: 301
> > Input tokens (cache write): 5048
> > Input tokens (cache read): 9191
> 
> > This failure matches my expectations, as without this change the result of the `format_messages_for_llm` will not have the cache_control property which is needed for prompt caching
> 
> I understand but it does have it, _when vision is enabled,_ it has it from `[message.model_dump() for message in messages]` code. This is the last message, 2 TextContent dicts: Screenshot 2024-09-09 at 20 19 02
> 
> So it must be that for you it didn't have `with_images` true at line 80. I'll have to try to replicate that. 🤔 You chose the model from the UI, and had Anthropic prefix, right? Which model exactly are you using?

Yes, confirmed that with images is being set to false. The root cause appears to be within the llm `_supports_vision` check.

The model being received is `anthropic/claude-3-5-sonnet-20240620`, which is not in litellms list (it's coded as "claude-3-5-sonnet-20240620". Basically the same problem that occurred in the other PR due to prefix mismatch.

https://raw.githubusercontent.com/BerriAI/litellm/main/model_prices_and_context_window.json
image

**Comment by amanape** [comment]
@enyst I didn't realize. I changed it because the previous model ID that was displayed (with a `.` separator) was either deprecated or incompatible. I've seen the lone Claude models and tested to see if they work after adding the `anthropic/` since it made it easier to parse the provider for the provider input. It all worked but I didn't test for images.

Its a little weird why LiteLLM returns and supports all these different variants 😭

**Comment by amanape** [comment]
Actually it is worth noting that I've done the same for the OpenAI provider... does the same issue persist for those models, too?

**Comment by tobitege** [comment]
On the server side we do need the provider prefix, like for OpenRouter an example would be "openrouter/anthropic/claude-3.5-sonnet". I guess we have to do extra parsing in the `_supports_vision()` method (llm.py) to be on the safe side.

**Comment by enyst** [comment]
@ColeMurray We wrote in the same time 😄.

You're right, that's it. OK, let's leave aside that issue for the moment. It means vision is broken for Anthropic models, so will need fixed, but that is not within the scope of this PR.

Back to the PR: the other issue here is that we merge TextContents into a single string per message, and that doesn't work well with prompt caching. Allow me to expand why. We have 4 checkpoints usually, the max allowed by Anthropic. The first is the system message, the second is initial user message, these work well and are permanently in place.

```
CodeActAgent LEVEL 0 LOCAL STEP 2 GLOBAL STEP 2
21:57:01 - openhands:INFO: llm.py:515 - Cost: 0.00 USD | Accumulated Cost: 0.03 USD
Input tokens: 4
Output tokens: 95
Input tokens (cache write): 268
Input tokens (cache read): 4610

```

Then third and fourth messages marked for caching are the last message in history, and the one before it. But the last message in history always has a [REMINDER added](https://github.com/All-Hands-AI/OpenHands/blob/c5e89be6dec3a6d1eb1bdc80a52046004457337c/agenthub/codeact_agent/codeact_agent.py#L288).
That means the last message with reminder is cached in one step, and it's _changed_ in the next step (because it doesn't include reminder anymore). This is a permanent cache miss, after the first few messages.
```

CodeActAgent LEVEL 0 LOCAL STEP 3 GLOBAL STEP 3

21:57:48 - openhands:INFO: llm.py:515 - Cost: 0.00 USD | Accumulated Cost: 0.04 USD
Input tokens: 4
Output tokens: 53
Input tokens (cache write): 380
Input tokens (cache read): 4610
...
CodeActAgent LEVEL 0 LOCAL STEP 7 GLOBAL STEP 7

21:59:04 - openhands:INFO: llm.py:515 - Cost: 0.01 USD | Accumulated Cost: 0.06 USD
Input tokens: 4
Output tokens: 124
Input tokens (cache write): 1798
Input tokens (cache read): 4610
```

This run will never cache more than 4610 tokens 😅. According to the links you posted, the messages must not change from one step to the other (of course):
> Changes that can break the cache include modifying any content

But the last message does always change, adding once the reminder and removing it next time. This works fine with Pydantic's serialization (if vision was enabled thing), because it doesn't merge the text content: it has 2 dicts for 2 TextContent objects, first is cached, the second is the reminder that will change. So that's fine, the first will be found in cache.

We can fix this in multiple ways, I suppose: instead of caching the last and previous-to-last message, we could cache the previous-to-last and the one before it; or, we could try to not merge the strings for text, serialize them as different dicts; or, we could hack the exclusion of the reminder text ... I don't favor the last, otherwise dunno, what do you think?

**Comment by enyst** [comment]
> On the server side we do need the provider prefix, like for OpenRouter an example would be "openrouter/anthropic/claude-3.5-sonnet". I guess we have to do extra parsing in the `_supports_vision()` method (llm.py) to be on the safe side.

That's an option, but just to clarify, I'm not talking about removing it everywhere, only those with "anthropic/" don't work for liteLLM. This exact prefix ("anthropic/") wasn't there before, and it doesn't affect the user or the UI for adding "openrouter/anthropic", liteLLM has the last, not the first.

**Comment by enyst** [comment]
> @enyst I didn't realize. I changed it because the previous model ID that was displayed (with a `.` separator) was either deprecated or incompatible. I've seen the lone Claude models and tested to see if they work after adding the `anthropic/` since it made it easier to parse the provider for the provider input. It all worked but I didn't test for images.
> 
> Its a little weird why LiteLLM returns and supports all these different variants 😭

Yes, frankly it feels like it should be added in `litellm.model_cost`. It does work to connect to the right thing, so... it's just not in the list for vision, even though they're the same models. Perhaps we can ask them.

**Comment by ColeMurray** [comment]
> @ColeMurray We wrote in the same time 😄.
> 
> You're right, that's it. OK, that is not within the scope of this PR.
> 
> Back to the PR: the other issue here is that we merge TextContents into a single string per message, and that doesn't work well with prompt caching. Allow me to expand why. We have 4 checkpoints usually, the max allowed by Anthropic. The first is the system message, the second is initial user message, these work well and are permanently in place.
> 
> ```
> CodeActAgent LEVEL 0 LOCAL STEP 2 GLOBAL STEP 2
> 21:57:01 - openhands:INFO: llm.py:515 - Cost: 0.00 USD | Accumulated Cost: 0.03 USD
> Input tokens: 4
> Output tokens: 95
> Input tokens (cache write): 268
> Input tokens (cache read): 4610
> ```
> 
> Then third and fourth messages marked for caching are the last message in history, and the one before it. But the last message in history always has a [REMINDER added](https://github.com/All-Hands-AI/OpenHands/blob/c5e89be6dec3a6d1eb1bdc80a52046004457337c/agenthub/codeact_agent/codeact_agent.py#L288). That means the last message with reminder is cached in one step, and it's _changed_ in the next step (because it doesn't include reminder anymore). This is a permanent cache miss:
> 
> ```
> 
> CodeActAgent LEVEL 0 LOCAL STEP 3 GLOBAL STEP 3
> 
> 21:57:48 - openhands:INFO: llm.py:515 - Cost: 0.00 USD | Accumulated Cost: 0.04 USD
> Input tokens: 4
> Output tokens: 53
> Input tokens (cache write): 380
> Input tokens (cache read): 4610
> ...
> CodeActAgent LEVEL 0 LOCAL STEP 7 GLOBAL STEP 7
> 
> 21:59:04 - openhands:INFO: llm.py:515 - Cost: 0.01 USD | Accumulated Cost: 0.06 USD
> Input tokens: 4
> Output tokens: 124
> Input tokens (cache write): 1798
> Input tokens (cache read): 4610
> ```
> 
> This run will never cache more than 4610 tokens 😅. According to the links you posted, the messages must not change from one step to the other (of course):
> 
> > Changes that can break the cache include modifying any content
> 
> But the last message does always change, adding once the reminder and removing it next time. This works fine with Pydantic's serialization (if vision was enabled thing), because it doesn't merge the text content: it has 2 dicts for 2 TextContent objects, first is cached, the second is the reminder that will change. So that's fine, the first will be found in cache.
> 
> We can fix this in multiple ways, I suppose: instead of caching the last and previous-to-last message, we could cache the previous-to-last and the one before it; or, we could try to not merge the strings for text, serialize them as different dicts, like `model_dump` is doing; or, we could hack the exclusion of the reminder text ... I don't favor the last, otherwise dunno, what do you think?

Confirming repro of the following due to the latest_user_message. I have this disabled locally to bypass this issue 😅 .

I think there's probably a larger design question here on how caching should be integrated into the codebase. The current implementation heavily focuses on Anthropic's implementation and feels a bit hacked in. If/when other providers add support, this will likely require a redesign.

Serializing them using the `model_dump` format makes the most sense to me, but I don't have full context on previous decisions for concating them together.


```
{
    'content': [
      {
        'type': 'text',
        'text': 'OBSERVATION:\nHello, World!\r\n\r\n[Python Interpreter: /openhands/poetry/openhands-ai-5O4_aCHf-py3.11/bin/python]\nopenhands@0632911a4cc9:/workspace $ \n[Command -1 finished with exit code 0]',
        'cache_control': {
          'type': 'ephemeral'
        }
      },
      {
        'type': 'text',
        'text': '\n\nENVIRONMENT REMINDER: You have 87 turns left to complete the task. When finished reply with <finish></finish>.'
      }
    ],
    'role': 'user'
  }
```

**Comment by enyst** [comment]
> I think there's probably a larger design question here on how caching should be integrated into the codebase. The current implementation heavily focuses on Anthropic's implementation and feels a bit hacked in. If/when other providers add support, this will likely require a redesign.

I agree, we will revisit it, I'm sure. Regardless how that refactoring will look like, I think still marking stuff from the agent (or some agent helper), then using a Message class for serialization makes the most sense to me too. 🤔 The agent/helper knows when history changes from a step to another, when the messages reach the LLM class it's too late. 🤷 

> Serializing them using the `model_dump` format makes the most sense to me, but I don't have full context on previous decisions for concating them together.

It's coming from this PR: https://github.com/All-Hands-AI/OpenHands/pull/3665
Cc: @tobitege That was a large PR which I don't fully understand and I guess changed after descriptions were written, because it doesn't seem to match, for example:
> This PR now sends content again as a single text if supports_prompt_caching is not active.

It doesn't do that though, it sends it if vision is not enabled. Maybe our quick fix could be exactly this: if any Message has prompt caching, just return `model_dump` like we do for vision?

> The class has the type attribute on the message level, which causes bad request errors for e.g. Gemini (on different providers) and Groq API.

It has `type` on the Content components of a message, TextContent and ImageContent. The stranger thing is that there's a Groq API screenshot in the thread which also shows `type` on both of these. But this error says it wants string:
> Example error that occured on e.g. Groq API with lama 3.1-70B (line breaks added for readability):
openai.BadRequestError: Error code: 400 - {'error': {'message': "'messages.0' : 
for 'role:system' the following must be satisfied[('messages.0.content' : value must be a string)]", 
'type': 'invalid_request_error'}}

Is their screenshot wrong or what am I missing? If the error message tells us the truth, it sounds like it can't be an array without `type` either, it has to be a string...Oh! the error is about system message, could it be that it needs a string only for the ***system*** message, while the rest, user messages, have dict arrays with `type`? 🤔 
I don't currently have Groq, but I'll do that soon.

@tobitege the PR fixes #3408. Can you please tell what was the fix for that? Was it fixed by the method `is_caching_prompt_active`, which was false for vertex?

**Comment by tobitege** [comment]
Yeh, while I was trying to reproduce the errors, I wasn't sure at first which feature the reason was and the pydantic class came in with the vision support, even before the caching, and I didn't fix the PR description afterwards. 😔 
I did also try to get the serialization without `type` in the TextContent, but had nothing but trouble with that and then decided to convert the messages back to just one content instead. As you mention the error message, it might be that the problem could've just been with the system message alone, with Groq I don't remember the specific error right now.

**Comment by enyst** [comment]
OK, I think we can fall back to `model_dump()` if prompt caching is in use, just like we do for vision. Thank you!

**Comment by ColeMurray** [comment]
Thanks for the context.

> OK, I think we can fall back to `model_dump()` if prompt caching is in use, just like we do for vision. Thank you!

I've updated the PR to add support for the discussed serialization when prompt caching is active. Given the identical behavior it might make sense in a future change to simplify to "serialize_messages:bool".

Separately, there's an action item on the frontend to fix the provider prefixing filed here https://github.com/All-Hands-AI/OpenHands/issues/3792

**Review by enyst — APPROVED** [review]
Thank you for doing this!

**Pull request merged**
